docs/versioning.md

```diff
@@ -0,0 +1,563 @@
+  or null version, in the object's metadata
+- send instructions to metadata backend to store, retrieve, update and delete
+  version metadata based on bucket versioning state and version metadata
+- encode version ID information to return in responses to requests, and decode
+  version IDs sent in requests
+
+The implementation of Bucket Versioning in S3 is described in this document in
+two main parts. The first section, "Implementation of Bucket Versioning in
+Metadata" describes the way versions are stored in metadata, and the metadata
+options for manipulating version metadata.
+
+The second section, "Implementation of Bucket Versioning in API" describes the
+way the metadata options are used in the API within S3 actions to create new
+versions, update their metadata, and delete them. The management of null
+versions and creation of delete markers are also described in this section.
+
+## Implementation of Bucket Versioning in Metadata
+
+As mentioned above, each version of an object is stored as a separate key in
+metadata. We use version identifiers as the suffix for the keys of the object
+versions, and a special version (the "Master Version") to represent the latest
+version.
+
+An example of what the metadata keys might look like for an object `foo/bar`
+with three versions:
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar\098506163554375999999PARIS  0.a430a1f85c6ec |
+| foo/bar\098506163554373999999PARIS  0.41b510cd0fdf8 |
+| foo/bar\098506163554373999998PARIS  0.f9b82c166f695 |
+
+The most recent version created is represented above in the key `foo/bar` and is
+the master version. This special version is described further in the subsection
+"Master Version". The subsection that immediately follows describes the metadata
+key and version ID format in further detail.
+
+### Version ID and Metadata Key Format
+
+The version ID is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3 before sending a response to a request. S3 also
+decodes the hexadecimal string received from a request before sending to
+metadata to retrieve a particular version.
+
+The format of a `version_id` is:
+`ts` `rep_group_id` `seq_id`
+where:
+
+- `ts`: is the combination of epoch and an increasing number
+- `rep_group_id`: is the name of deployment(s) considered one unit used for
+   replication
+- `seq_id`: is supposed to be a unique value based on metadata information.
+
+The format of a key in metadata for a version is:
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the format described above as metadata orders keys
+  alphabetically
+
+An example of a key in metadata: `foo\01234567890000777PARIS 1234.123456`
```

Comment:
I think the space is actually due to padding the rg_token. The padding character for the rg_token is `' '`.